psi4/src/psi4/dlpno/mp2.cc

```diff
@@ -0,0 +1,1424 @@
+ */
+void copy_flat_mats(SharedMatrix flat, std::vector<SharedMatrix>& mat_list) {
+    double** flatp = flat->pointer();
+    size_t flat_ind = 0;
+    for (SharedMatrix mat : mat_list) {
+        if (mat->size() == 0) continue;
+        ::memcpy(mat->pointer()[0], &flatp[0][flat_ind], sizeof(double) * mat->size());
+        flat_ind += mat->size();
+    }
+}
+
+/* Args: orthonormal orbitals C (ao x mo) and fock matrix F (ao x ao)
+ * Return: transformation matrix X (mo x mo) and energy vector e (mo)
+ *
+ * CX are canonical orbitals (i.e. F(CX) = e(CX))
+ */
+std::pair<SharedMatrix, SharedVector> canonicalizer(SharedMatrix C, SharedMatrix F) {
+    SharedMatrix X = std::make_shared<Matrix>("eigenvectors", C->colspi(0), C->colspi(0));
+    SharedVector e = std::make_shared<Vector>("eigenvalues", C->colspi(0));
+
+    auto temp = linalg::triplet(C, F, C, true, false, false);
+    temp->diagonalize(X, e, descending);
+
+    return std::make_pair(X, e);
+}
+
+/* Args: overlap matrix S (mo x mo), fock matrix F (mo x mo)
+ * Return: transformation matrix X (mo x mo_new) and energy vector e (mo_new)
+ *
+ * CX are orthonormal orbitals (i.e. S_ao (CX) = CX) and also canonical (i.e. F_ao (CX) = e(CX))
+ * linear dependencies are removed with keyword S_CUT, so (mo_new <= mo)
+ */
+std::pair<SharedMatrix, SharedVector> DLPNOMP2::orthocanonicalizer(SharedMatrix S, SharedMatrix F) {
+    BasisSetOrthogonalization orthog(BasisSetOrthogonalization::PartialCholesky, S, 0.0, options_.get_double("S_CUT"),
+                                     0);
+    auto X = orthog.basis_to_orthog_basis();
+
+    int nmo_initial = X->rowspi(0);
+    int nmo_final = X->colspi(0);
+
+    auto U = std::make_shared<Matrix>("eigenvectors", nmo_final, nmo_final);
+    auto e = std::make_shared<Vector>("eigenvalues", nmo_final);
+
+    auto F_orth = linalg::triplet(X, F, X, true, false, false);
+    F_orth->diagonalize(U, e, descending);
+
+    X = linalg::doublet(X, U, false, false);
+
+    return std::make_pair(X, e);
+}
+
+void DLPNOMP2::compute_overlap_ints() {
+    int nbf = basisset_->nbf();
+    int naocc = C_lmo_->colspi(0);
+    int npao = C_pao_->colspi(0);  // same as nbf
+
+    timer_on("Construct Grid");
+    auto grid = DFTGrid(molecule_, basisset_, options_);
+    timer_off("Construct Grid");
+
+    size_t nthread = 1;
+#ifdef _OPENMP
+    nthread = omp_get_max_threads();
+#endif
+    std::vector<std::shared_ptr<BasisFunctions>> point_funcs(nthread);
+    std::vector<SharedMatrix> DOI_ij_temps(nthread);
+    std::vector<SharedMatrix> DOI_iu_temps(nthread);
+    for (size_t thread = 0; thread < nthread; thread++) {
+        point_funcs[thread] = std::make_shared<BasisFunctions>(basisset_, grid.max_points(), nbf);
+        DOI_ij_temps[thread] = std::make_shared<Matrix>("(i,j) Differential Overlap Integrals", naocc, naocc);
+        DOI_iu_temps[thread] = std::make_shared<Matrix>("(i,u) Differential Overlap Integrals", naocc, nbf);
```

Comment:
These look to me like more completely unnecessary pointers. Are these needed for parallelization, for some reason?

Reply:
We need `nthread` of these items so that each thread has one of its own. I changed the `DOI_ij_temps` and `DOI_iu_temps` to use regular `Matrix` objects. make a `BasisFunctions` object without a shared pointer causes a strange compilation error, so I left it as is